File: backend/services/stigmer-server/pkg/domain/workflowexecution/controller/create_execution_context_step.go
package workflowexecution

import (
	"fmt"

	"github.com/rs/zerolog/log"
	environmentv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/environment/v1"
	executioncontextv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/executioncontext/v1"
	workflowv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflow/v1"
	workflowexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/workflowexecution/v1"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource"
	"github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/commons/apiresource/apiresourcekind"
	"github.com/stigmer/stigmer/backend/libs/go/envmerge"
	"github.com/stigmer/stigmer/backend/libs/go/grpc/request/pipeline"
	"github.com/stigmer/stigmer/backend/libs/go/store"
	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/downstream/environment"
	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/downstream/executioncontext"
	"github.com/stigmer/stigmer/backend/services/stigmer-server/pkg/downstream/workflowinstance"
)

// createExecutionContextStep builds and persists an ExecutionContext with a fully-merged
// environment for the workflow execution.
//
// Resolution chain:
//   - Resolve workflow_instance_id from pipeline context (ResolvedWorkflowInstanceIDKey) or execution spec
//   - workflowInstanceClient.Get -> WorkflowInstance.env_refs + workflow_id
//   - store.GetResource(workflow_id) -> Workflow.env_spec (follows WE controller's "same service" store-access pattern)
//
// Merge priority (lowest to highest):
//  1. WorkflowInstance.env_refs resolved via environmentClient (in order)
//  2. WorkflowExecution.spec.runtime_env (execution-time overrides)
type createExecutionContextStep struct {
	store                  store.Store
	workflowInstanceClient *workflowinstance.Client
	environmentClient      *environment.Client
	executionCtxClient     *executioncontext.Client
}

func (c *WorkflowExecutionController) newCreateExecutionContextStep() *createExecutionContextStep {
	return &createExecutionContextStep{
		store:                  c.store,
		workflowInstanceClient: c.workflowInstanceClient,
		environmentClient:      c.environmentClient,
		executionCtxClient:     c.executionContextClient,
	}
}

func (s *createExecutionContextStep) Name() string {
	return "CreateExecutionContext"
}

func (s *createExecutionContextStep) Execute(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecution]) error {
	execution := ctx.NewState()
	executionID := execution.GetMetadata().GetId()
	executionOrg := execution.GetMetadata().GetOrg()

	if s.workflowInstanceClient == nil || s.environmentClient == nil || s.executionCtxClient == nil {
		log.Warn().
			Str("execution_id", executionID).
			Msg("ExecutionContext clients not available, skipping execution context creation")
		return nil
	}

	log.Debug().
		Str("execution_id", executionID).
		Msg("Creating execution context with merged environment")

	// 1. Resolve workflow_instance_id
	workflowInstanceID := s.resolveWorkflowInstanceID(ctx)
	if workflowInstanceID == "" {
		return fmt.Errorf("workflow_instance_id not resolved from context or execution spec")
	}

	log.Debug().
		Str("execution_id", executionID).
		Str("workflow_instance_id", workflowInstanceID).
		Msg("Resolved workflow instance ID")

	// 2. Load WorkflowInstance to get env_refs and workflow_id
	instance, err := s.workflowInstanceClient.Get(ctx.Context(), workflowInstanceID)
	if err != nil {
		return fmt.Errorf("load workflow instance %s: %w", workflowInstanceID, err)
	}

	workflowID := instance.GetSpec().GetWorkflowId()

	// 3. Load Workflow from store to get env declarations (follows WE controller's "same service" pattern)
	workflow := &workflowv1.Workflow{}
	if err := s.store.GetResource(ctx.Context(), apiresourcekind.ApiResourceKind_workflow, workflowID, workflow); err != nil {
		return fmt.Errorf("load workflow %s: %w", workflowID, err)
	}

	// 4. Resolve environments from instance env_refs
	environments, err := s.resolveEnvironments(ctx, instance.GetSpec().GetEnvironmentRefs())
	if err != nil {
		return err
	}

	// 5. Merge all layers
	merged := envmerge.MergeEnvironmentLayers(
		environments,
		execution.GetSpec().GetRuntimeEnv(),
	)

	// 6. Filter merged env vars by workflow env declarations (least-privilege whitelist).
	// Workflows only receive variables they explicitly declared. If env is
	// nil or empty, all vars pass through for backward compatibility.
	workflowEnvDecls := workflow.GetSpec().GetEnv()
	filtered, excludedKeys := envmerge.FilterByDeclaredKeys(merged, workflowEnvDecls)
	if len(excludedKeys) > 0 {
		log.Warn().
			Str("execution_id", executionID).
			Str("workflow_id", workflowID).
			Strs("excluded_keys", excludedKeys).
			Msg("Filtered env vars not declared in workflow env")
	}

	// 6.1 Validate that all required (non-optional) declared env vars are
	// present after merging and filtering. Workflows access MCP servers
	// through agent tasks whose executions are validated separately, so
	// this primarily catches missing workflow-level required vars.
	if missingRequired := envmerge.ValidateRequiredKeys(filtered, workflowEnvDecls); len(missingRequired) > 0 {
		log.Warn().
			Str("execution_id", executionID).
			Str("workflow_id", workflowID).
			Strs("missing_required", missingRequired).
			Msg("Required env vars missing after environment merge — execution may fail")
	}

	log.Info().
		Str("execution_id", executionID).
		Int("merged_count", len(merged)).
		Int("filtered_count", len(filtered)).
		Int("env_refs_count", len(instance.GetSpec().GetEnvironmentRefs())).
		Msg("Merged environment layers for execution context")

	// 7. Build and persist ExecutionContext
	ec := &executioncontextv1.ExecutionContext{
		ApiVersion: "agentic.stigmer.ai/v1",
		Kind:       "ExecutionContext",
		Metadata: &apiresource.ApiResourceMetadata{
			Name: fmt.Sprintf("exec-ctx-%s", executionID),
			Org:  executionOrg,
		},
		Spec: &executioncontextv1.ExecutionContextSpec{
			ExecutionId: executionID,
			Data:        filtered,
		},
	}

	created, err := s.executionCtxClient.Create(ctx.Context(), ec)
	if err != nil {
		return fmt.Errorf("create execution context for %s: %w", executionID, err)
	}

	log.Info().
		Str("execution_context_id", created.GetMetadata().GetId()).
		Str("execution_id", executionID).
		Int("data_entries", len(filtered)).
		Msg("Successfully created execution context")

	// 8. Clear runtime_env from the execution now that it has been consumed.
	// runtime_env is a transient creation-time input; its contents are now
	// materialized in the ExecutionContext. Clearing it ensures secrets never
	// appear in the persisted execution or in Temporal workflow history.
	if execution.GetSpec() != nil && len(execution.GetSpec().GetRuntimeEnv()) > 0 {
		log.Debug().
			Str("execution_id", executionID).
			Int("cleared_entries", len(execution.GetSpec().GetRuntimeEnv())).
			Msg("Clearing runtime_env from execution (consumed into ExecutionContext)")

		execution.Spec.RuntimeEnv = nil
		ctx.SetNewState(execution)
	}

	return nil
}

// resolveWorkflowInstanceID determines the workflow_instance_id from pipeline context
// (set by createDefaultInstanceIfNeededStep) or from the execution spec.
func (s *createExecutionContextStep) resolveWorkflowInstanceID(ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecution]) string {
	// The createDefaultInstanceIfNeededStep always sets workflow_instance_id on execution.spec
	// regardless of whether it was user-provided or auto-resolved from default instance.
	execution := ctx.NewState()
	return execution.GetSpec().GetWorkflowInstanceId()
}

// resolveEnvironments fetches each referenced Environment resource in order.
func (s *createExecutionContextStep) resolveEnvironments(
	ctx *pipeline.RequestContext[*workflowexecutionv1.WorkflowExecution],
	refs []*apiresource.ApiResourceReference,
) ([]*environmentv1.Environment, error) {
	if len(refs) == 0 {
		return nil, nil
	}

	environments := make([]*environmentv1.Environment, 0, len(refs))
	for _, ref := range refs {
		env, err := s.environmentClient.GetByReference(ctx.Context(), ref)
		if err != nil {
			return nil, fmt.Errorf("resolve environment ref (org=%s, slug=%s): %w",
				ref.GetOrg(), ref.GetSlug(), err)
		}
		environments = append(environments, env)
	}

	log.Debug().
		Int("count", len(environments)).
		Msg("Resolved environment references")

	return environments, nil
}

File: backend/services/stigmer-server/src/domain/workflowexecution/create-execution-context-step.ts
/**
 * CreateExecutionContext — ports create_execution_context_step.go: builds
 * and persists the ExecutionContext carrying the fully-merged environment
 * for the run, then strips spec.runtime_env so secrets never reach the
 * persisted execution or Temporal history.
 *
 * Resolution chain: execution.spec.workflow_instance_id (always set by
 * CreateDefaultInstanceIfNeeded) → instance.environment_refs resolved
 * through the environment RuntimeResolutionService (decrypted — the RPC
 * surface redacts, oss#405) → envmerge with spec.runtime_env as the top
 * layer → least-privilege filter against the workflow's env declarations
 * (undeclared keys warn + drop; empty declarations pass everything for
 * backward compatibility) → required-key validation (warn-only) →
 * ExecutionContext create through the in-process client.
 *
 * Go skips the whole step when its late-injected deps are nil; the TS
 * composition root wires them unconditionally, so that arm is
 * structurally unreachable here and deliberately not modeled (the
 * loud-boot doctrine: a missing dependency is a wiring bug, not a mode).
 *
 * The recover pipeline's RecreateExecutionContext twin lives in
 * lifecycle.ts — its failure posture differs (degrade gracefully).
 */
import { create } from "@bufbuild/protobuf";
import { ConnectError } from "@connectrpc/connect";

import type { Environment } from "@stigmer/protos/ai/stigmer/agentic/environment/v1/api_pb";
import type { ExecutionContext } from "@stigmer/protos/ai/stigmer/agentic/executioncontext/v1/api_pb";
import { ExecutionContextSchema } from "@stigmer/protos/ai/stigmer/agentic/executioncontext/v1/api_pb";
import { WorkflowSchema } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/api_pb";
import type { Workflow } from "@stigmer/protos/ai/stigmer/agentic/workflow/v1/api_pb";
import type { WorkflowInstance } from "@stigmer/protos/ai/stigmer/agentic/workflowinstance/v1/api_pb";
import { WorkflowExecutionSchema } from "@stigmer/protos/ai/stigmer/agentic/workflowexecution/v1/api_pb";
import type { ApiResourceReference } from "@stigmer/protos/ai/stigmer/commons/apiresource/io_pb";
import { ApiResourceKind } from "@stigmer/protos/ai/stigmer/commons/apiresource/apiresourcekind/api_resource_kind_pb";

import type { Logger } from "../../boot/logger.js";
import type { RuntimeResolutionService } from "../../domain/environment/resolution/resolution.js";
import {
  filterByDeclaredKeys,
  mergeEnvironmentLayers,
  validateRequiredKeys,
} from "../../envmerge/envmerge.js";
import {
  goWrappedStatusError,
  internalError,
} from "../../pipeline/errors.js";
import type { PipelineStep } from "../../pipeline/pipeline.js";
import type { Store } from "../../store/interface.js";

/**
 * The narrow workflowinstance READ edge (Go workflowInstanceClient.Get)
 * — consumer-defined, satisfied by the in-process query client.
 */
export interface ExecutionWorkflowInstanceLoader {
  get(instanceId: string): Promise<WorkflowInstance>;
}
export type ExecutionWorkflowInstanceLoaderProvider =
  () => ExecutionWorkflowInstanceLoader;

/** The narrow executioncontext CREATE edge (Go executionCtxClient.Create). */
export interface WorkflowExecutionContextCreator {
  create(executionContext: ExecutionContext): Promise<ExecutionContext>;
}
export type WorkflowExecutionContextCreatorProvider =
  () => WorkflowExecutionContextCreator;

export interface WorkflowExecutionContextBuilderDeps {
  readonly store: Store;
  readonly logger: Logger;
  readonly workflowInstanceLoader: ExecutionWorkflowInstanceLoaderProvider;
  /** The decrypt-for-execution path (oss#405); a direct service, no RPC. */
  readonly environmentResolution: RuntimeResolutionService;
  readonly executionContextCreator: WorkflowExecutionContextCreatorProvider;
}

export function newCreateExecutionContextStep(
  deps: WorkflowExecutionContextBuilderDeps,
): PipelineStep<typeof WorkflowExecutionSchema> {
  return {
    name: "CreateExecutionContext",
    async execute(ctx) {
      const execution = ctx.newState;
      const executionId = execution.metadata?.id ?? "";
      const executionOrg = execution.metadata?.org ?? "";

      // CreateDefaultInstanceIfNeeded always stamps this, user-provided
      // or auto-resolved.
      const workflowInstanceId = execution.spec?.workflowInstanceId ?? "";
      if (workflowInstanceId === "") {
        throw internalError(
          new Error(
            "workflow_instance_id not resolved from context or execution spec",
          ),
          "workflow_instance_id not resolved from context or execution spec",
        );
      }

      let instance: WorkflowInstance;
      try {
        instance = await deps.workflowInstanceLoader().get(workflowInstanceId);
      } catch (error) {
        if (error instanceof ConnectError) {
          throw goWrappedStatusError(
            `load workflow instance ${workflowInstanceId}`,
            error,
          );
        }
        throw internalError(
          error,
          `load workflow instance ${workflowInstanceId}`,
        );
      }

      const workflowId = instance.spec?.workflowId ?? "";
      let workflow: Workflow;
      try {
        workflow = await deps.store.getResource(
          ApiResourceKind.workflow,
          workflowId,
          WorkflowSchema,
        );
      } catch (error) {
        throw internalError(error, `load workflow ${workflowId}`);
      }

      const environments = await resolveEnvironments(
        deps,
        instance.spec?.environmentRefs ?? [],
      );

      const merged = mergeEnvironmentLayers(
        environments,
        execution.spec?.runtimeEnv ?? {},
      );

      // Least-privilege whitelist: workflows only receive declared vars.
      const workflowEnvDecls = workflow.spec?.env ?? {};
      const { filtered, excludedKeys } = filterByDeclaredKeys(
        merged,
        workflowEnvDecls,
      );
      if (excludedKeys.length > 0) {
        deps.logger.warn("Filtered env vars not declared in workflow env", {
          executionId,
          workflowId,
          excludedKeys,
        });
      }

      const missingRequired = validateRequiredKeys(filtered, workflowEnvDecls);
      if (missingRequired.length > 0) {
        deps.logger.warn(
          "Required env vars missing after environment merge — execution may fail",
          { executionId, workflowId, missingRequired },
        );
      }

      const executionContext = create(ExecutionContextSchema, {
        apiVersion: "agentic.stigmer.ai/v1",
        kind: "ExecutionContext",
        metadata: {
          name: `exec-ctx-${executionId}`,
          org: executionOrg,
        },
        spec: {
          executionId,
          data: Object.fromEntries(filtered),
        },
      });

      try {
        const created = await deps
          .executionContextCreator()
          .create(executionContext);
        deps.logger.info("Successfully created execution context", {
          executionContextId: created.metadata?.id ?? "",
          executionId,
          dataEntries: filtered.size,
        });
      } catch (error) {
        if (error instanceof ConnectError) {
          throw goWrappedStatusError(
            `create execution context for ${executionId}`,
            error,
          );
        }
        throw internalError(
          error,
          `create execution context for ${executionId}`,
        );
      }

      // runtime_env is a transient creation-time input, now materialized
      // in the ExecutionContext; clearing it keeps secrets out of the
      // persisted execution and Temporal history.
      if (
        execution.spec !== undefined &&
        Object.keys(execution.spec.runtimeEnv).length > 0
      ) {
        execution.spec.runtimeEnv = {};
      }
    },
  };
}

/**
 * Fetches each referenced Environment in order through the runtime
 * resolution service (decrypted values; the RPC surface redacts,
 * oss#405). A failed ref fails the create — Go wraps with the ref
 * coordinates.
 */
async function resolveEnvironments(
  deps: WorkflowExecutionContextBuilderDeps,
  refs: ApiResourceReference[],
): Promise<Environment[]> {
  if (refs.length === 0) {
    return [];
  }
  const environments: Environment[] = [];
  for (const ref of refs) {
    try {
      environments.push(await deps.environmentResolution.resolveByReference(ref));
    } catch (error) {
      const prefix = `resolve environment ref (org=${ref.org}, slug=${ref.slug})`;
      if (error instanceof ConnectError) {
        throw goWrappedStatusError(prefix, error);
      }
      throw internalError(error, prefix);
    }
  }
  return environments;
}
